### PR TITLE
Support JMH for scala 2.12 #295

### PR DESCRIPTION
**Pull request by dkomanov**
Use reflection-based generator (RFGeneratorSource) instead of ASM-based, it does support Java8 byte-code.

Also version of JMH is upgraded.

### CODE REVIEW

File: src/scala/io/bazel/rules_scala/jmh_support/BenchmarkGenerator.scala [hunk 1]
```diff
@@ -47,17 +55,18 @@ object BenchmarkGenerator {
...
-        "Usage: BenchmarkGenerator INPUT_JAR RESULT_JAR RESOURCE_JAR [CLASSPATH_ELEMENT] [CLASSPATH_ELEMENT...]"
+        "Usage: BenchmarkGenerator GENERATOR_TYPE INPUT_JAR RESULT_JAR RESOURCE_JAR [CLASSPATH_ELEMENT] [CLASSPATH_ELEMENT...]"
       )
       System.exit(1)
     }
     val fs = FileSystems.getDefault
 
     BenchmarkGeneratorArgs(
-      fs.getPath(argv(0)),
+      if ("asm".equalsIgnoreCase(argv(0))) AsmGenerator else ReflectionGenerator,
```
Comment (johnynek) [inline src/scala/io/bazel/rules_scala/jmh_support/BenchmarkGenerator.scala]: can we error if it is not asm or reflection?

File: src/scala/io/bazel/rules_scala/jmh_support/BenchmarkGenerator.scala [hunk 2]
```diff
@@ -131,17 +141,33 @@ object BenchmarkGenerator {
...
+        val source: GeneratorSource = generatorType match {
+          case AsmGenerator =>
+            val generatorSource = new ASMGeneratorSource
+            try {
+              generatorSource.processClasses(collectClassesFromJar(benchmarkJarPath).map(_.toFile).asJavaCollection)
+            } catch {
+              case _: ArrayIndexOutOfBoundsException =>
+                // this would fail due to https://github.com/bazelbuild/rules_scala/issues/295
+                // let's throw a useful message instead
+                sys.error("jmh in rules_scala doesn't work with Java 8 bytecode: https://github.com/bazelbuild/rules_scala/issues/295")
```
Comment (johnynek) [inline src/scala/io/bazel/rules_scala/jmh_support/BenchmarkGenerator.scala]: can we change the error to say asm mode does not work with java 8 bytecode and instead suggest using reflection?

**Comment by johnynek** [comment]
Thank you so much for tackling this! This will be a big help.

I think the 2.12 branch is a bit out of date sadly. For instance, I think it may have an out of date `.bazelci/presubmit.yml`

Can you make this as a PR against master and then we can cherry-pick it and re-enable the jmh test?

If you do this, we can probably merge very quickly.

**Comment by dkomanov** [comment]
Closing in favor of #465

**Pull request closed without merging**